src/toxcore/tcp/links.rs

```diff
@@ -0,0 +1,300 @@
+    /// The status of the link
+    pub status: LinkStatus,
+}
+
+impl Link {
+    /// Create a new link with the given PK and with status = Registered
+    fn new(pk: PublicKey) -> Link {
+        Link {
+            pk,
+            status: LinkStatus::Registered,
+        }
+    }
+    /// Change status to Registered
+    fn downgrade(&mut self) {
+        self.status = LinkStatus::Registered;
+    }
+    /// Change status to Online
+    fn upgrade(&mut self) {
+        self.status = LinkStatus::Online;
+    }
+}
+
+/// The structure contains `MAX_LINKS_N` of links
+pub struct Links {
+    /** An array of Link with `MAX_LINKS_N` item
+    None means there is a hole and you can allocate the `Link`
+    and respond the id with RouteResponse. Packets like `Data`
+    use connection_id as an index for links.
+    */
+    links: [Option<Link>; MAX_LINKS_N],
+    /// Map PK -> id to index links with O(1)
+    pk_to_id: HashMap<PublicKey, u8>,
+}
+
+impl Links {
+    /// Create empty `Links`
+    pub fn new() -> Links {
+        Links {
+            links: [None; 240],
+            pk_to_id: HashMap::new()
+        }
+    }
+    /** Try to find a hole inside `links` and insert a new `Link` to the given PK
+    Return `Some(id)` if there was a `Link` or if there is a room for a new `Link`
+    Return `None` if there is no room for a new `Link`
+    */
+    pub fn insert(&mut self, pk: &PublicKey) -> Option<u8> {
+        let possible_index = { self.pk_to_id.get(pk).cloned() };
+        match possible_index {
+            Some(index) => Some(index), // already inserted
+            None => {
+                if let Some(index) = self.links.iter().position(|link| link.is_none()) {
+                    let link = Link::new(*pk);
+                    self.links[index] = Some(link);
+                    self.pk_to_id.insert(*pk, index as u8);
+                    Some(index as u8)
+                } else {
+                    // no enough room for a link
+                    None
+                }
+            }
+        }
+    }
+    /// Get `Link` by id
+    pub fn by_id(&self, index: u8) -> Option<&Link> {
+        let index = index as usize;
+        if index < MAX_LINKS_N {
+            self.links[index].as_ref()
+        } else {
+            None
+        }
+    }
+    /// Get index of the link by PK
+    pub fn id_by_pk(&self, pk: &PublicKey) -> Option<u8> {
+        self.pk_to_id.get(pk).cloned()
+    }
+    /// Takes the link out of the links, leaving a None in its place
+    pub fn take(&mut self, index: u8) -> Option<Link> {
+        let index = index as usize;
+        if index < MAX_LINKS_N {
+            if let Some(link) = self.links[index].take() {
+                self.pk_to_id.remove(&link.pk);
+                Some(link)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+    /// Call Links::downgrade on the `Link` by id
+    pub fn downgrade(&mut self, index: u8) {
```

Comment:
How about returning `bool` if link by the index exists? It could be handy here: https://github.com/tox-rs/tox/blob/4d9ce78351e7ad78ae65b0e177b25ae1676c0306/src/toxcore/tcp/client/client.rs#L198-L206

Reply:
What logic should be behind this bool?

Reply:
What if the link exists and with status=Registered?

Reply:
This method downgrades state. So it will be always with status `Registered` if it exists. So just return true. And the same with `upgrade` method and `Online` status.

Reply:
What about `try_insert`/`try_downgrade`/`try_upgrade`?

Reply:
You mean rename it? Because it seems no reason to have `upgrade` and `try_upgrade` at the same time.

Reply:
Yes I mean rename them.

Reply:
Shall I or is there any reason to do so?

Reply:
Done